mlflow/metrics/metric_definitions.py

```diff
@@ -373,6 +373,94 @@ def _fn(predictions, targets):
     return _fn
 
 
+def _expand_duplicate_retrieved_docs(predictions, targets):
+    counter = {}
+    expanded_predictions = []
+    expanded_targets = targets.copy()
+    for doc_id in predictions:
+        if doc_id not in counter:
+            counter[doc_id] = 1
+            expanded_predictions.append(doc_id)
+        else:
+            counter[doc_id] += 1
+            new_doc_id = (
+                f"{doc_id}_bc574ae_{counter[doc_id]}"  # adding a random string to avoid collisions
+            )
+            expanded_predictions.append(new_doc_id)
+            expanded_targets.add(new_doc_id)
+    return expanded_predictions, expanded_targets
+
+
+def _prepare_row_for_ndcg(predictions, targets):
+    """Prepare data one row from predictions and targets to y_score, y_true for ndcg calculation.
+
+    Args:
+        predictions: A list of strings of at most k doc IDs retrieved.
+        targets: A list of strings of ground-truth doc IDs.
+
+    Returns:
+        y_true : ndarray of shape (1, n_docs) Representing the ground-truth relevant docs.
+            n_docs is the number of unique docs in union of predictions and targets.
+        y_score : ndarray of shape (1, n_docs) Representing the retrieved docs.
+            n_docs is the number of unique docs in union of predictions and targets.
+    """
+    eps = 1e-6
+    targets = set(targets)
+
+    # support predictions contain duplicate doc ID
+    predictions, targets = _expand_duplicate_retrieved_docs(predictions, targets)
+
+    all_docs = targets.union(predictions)
+    doc_id_to_index = {doc_id: i for i, doc_id in enumerate(all_docs)}
+    n_labels = max(len(doc_id_to_index), 2)  # sklearn.metrics.ndcg_score requires at least 2 labels
+    y_true = np.zeros((1, n_labels), dtype=np.float32)
+    y_score = np.zeros((1, n_labels), dtype=np.float32)
+    for i, doc_id in enumerate(predictions):
+        # "1 - i * eps" means we assign higher score to docs that are ranked higher,
```

Comment:
How do we know if the docs are ranked higher? We don't store some count anywhere as far as I can tell. Let me know if I am missing something

Reply:
1. The retrieved rank of the doc is represented by the order in the list in  `predictions` column.
2. The ground truth does not distinguish the rank of the docs. Our ndcg score views all documents in the `targets` column as an unordered set since we use binary relevance (for simplicity).